libs/api/ebpf_api.cpp

```diff
@@ -673,3 +676,401 @@ ebpf_api_map_info_free(_In_ const uint16_t map_count, _In_count_(map_count) cons
+    map->map_defintion.type = (ebpf_map_type_t)map_cache.ebpf_map_descriptor.type;
+    map->map_defintion.key_size = map_cache.ebpf_map_descriptor.key_size;
+    map->map_defintion.value_size = map_cache.ebpf_map_descriptor.value_size;
+    map->map_defintion.max_entries = map_cache.ebpf_map_descriptor.max_entries;
+    map->pinned = false;
+    map->pin_path = nullptr;
+}
+
+static ebpf_result_t
+_initialize_ebpf_object_from_elf(
+    const char* file_name,
+    const ebpf_program_type_t* expected_program_type,
+    const ebpf_attach_type_t* expected_attach_type,
+    ebpf_object_t& object,
+    const char** error_message) noexcept
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    set_global_program_and_attach_type(expected_program_type, expected_attach_type);
+
+    ebpf_verifier_options_t verifier_options{false, false, false, false, false};
+    result =
+        load_byte_code(file_name, nullptr, &verifier_options, &object.programs, &object.programs_count, error_message);
+    if (result != EBPF_SUCCESS) {
+        goto Exit;
+    }
+
+    object.file_name = allocate_string(file_name);
+    if (object.file_name == nullptr) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+
+    try {
+        auto map_descriptors = get_all_map_descriptors();
+        for (const auto& descriptor : map_descriptors) {
+            ebpf_map_t* map = (ebpf_map_t*)calloc(1, sizeof(ebpf_map_t));
+            if (map == nullptr) {
+                result = EBPF_NO_MEMORY;
+                goto Exit;
+            }
+
+            _initialize_map(map, &object, descriptor);
+            ebpf_list_insert_tail(&object.maps, &map->list_entry);
+            object.maps_count++;
+        }
+    } catch (const std::bad_alloc&) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    } catch (...) {
+        result = EBPF_FAILED;
+        goto Exit;
+    }
+
+Exit:
+    if (result != EBPF_SUCCESS) {
+        clean_up_ebpf_programs(object.programs);
+    }
+    return result;
+}
+
+_Success_(result == EBPF_SUCCESS) ebpf_result_t ebpf_program_load(
+    _In_z_ const char* file_name,
+    _In_opt_ const ebpf_program_type_t* program_type,
+    _In_opt_ const ebpf_attach_type_t* attach_type,
+    _Outptr_ struct ebpf_object** object,
+    _Out_ fd_t* program_fd,
+    _Outptr_opt_result_z_ const char** log_buffer)
+{
+    ebpf_object_t* new_object = nullptr;
+    ebpf_protocol_buffer_t request_buffer;
+    uint32_t error_message_size = 0;
+    std::vector<uintptr_t> handles;
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_program_load_info load_info = {0};
+    std::vector<fd_handle_map> handle_map;
+    ebpf_list_entry_t* entry;
+    ebpf_program_t* program;
+    ebpf_map_t* map;
+
+    if (file_name == nullptr || object == nullptr || program_fd == nullptr || log_buffer == nullptr) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    // If custom attach type is provided, then custom program
+    // type should also be provided.
```

Comment:
Why?  You should be able to look up the program type from the attach type, no?

Reply:
The intent here is that caller either provides both, or provides none. The caller can provide program_type in this API, and choose to not provide attach type. The attach API again provides another chance to provide attach type.

Reply:
I'm asking why that is the intent... is that what libbpf requires?

Reply:
I checked and libbpf also does the same (set the provided program type and attach type only when the program type is provided)

Also, since this API is for loading program (attach type is actually not used while loading the program, it is saved in the ebpf_program object for any future attach), and caller is choosing to override the default program/attach types (default = the values that can be derived from section name), I think it is intuitive that the caller should provide expected program type also, instead of only providing expected attach type?

Reply:
Ok as long as libbpf does the same, all good.